File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left = [-1] * len(height)
        right = [-1] * len(height)
        curr_max_left = -1
        curr_max_right = -1
        N = len(height)
        res = 0
        for i in range(N):
            curr_max_left = max(curr_max_left,height[i])
            curr_max_right = max(curr_max_right,height[N - i - 1])
            left[i] = curr_max_left
            right[N-i-1] = curr_max_right
        for i in range(N):
            res += abs(min(left[i],right[i])-height[i])
        return res

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int [] leftPeak = new int[height.length];
        int [] rightPeak = new int[height.length];
        int res = 0;
        int currLeftPeak = height[0];
        int currRightPeak = height[height.length - 1];
        for(int i = 0;i<height.length;i++){
            currLeftPeak = Math.max(height[i],currLeftPeak);
            leftPeak[i] = currLeftPeak;
        }
        for(int i = height.length - 1;i>=0;i--){
            currRightPeak = Math.max(height[i],currRightPeak);
            rightPeak[i] = currRightPeak;
        }
        for(int i = 0;i<height.length;i++){
            res += Math.min(rightPeak[i],leftPeak[i]) - height[i];
        }
        return(res); 
    }
}